python/tk_framework_desktopserver/shotgun/scripts/cache_commands.py

```diff
@@ -0,0 +1,185 @@
+    toolkit_mgr.plugin_id = "shotgun_toolkit_menu_caching"
+    toolkit_mgr.base_configuration = base_configuration
+
+    for pc in data["pipeline_configs"]:
+        pc["project"] = pc.get("project", project_entity)
+    import pprint
+    pprint.pprint(data["pipeline_configs"])
+    pcs = toolkit_mgr.sort_and_filter_configuration_entities(
+        project=project_entity,
+        entities=data["pipeline_configs"],
+    ) or [dict(id=None)]
+
+    for pc in pcs:
+        logger, log_handler = get_sgtk_logger(sgtk)
+        pprint.pprint(hash_data)
+        lookup_hash = hash_data[pc["id"]]["lookup_hash"]
+        contents_hash = hash_data[pc["id"]]["contents_hash"]
+
+        toolkit_mgr.pipeline_configuration = pc["id"]
+        engine = toolkit_mgr.bootstrap_engine(engine_name, entity=entity)
+        pc_descriptor = toolkit_mgr.resolve_descriptor(project_entity)
+
+        # Clean up the pre-bootstrap logger since we can use the engine's
+        # logger from here on out.
+        sgtk.LogManager().root_logger.removeHandler(log_handler)
+        engine.logger.debug("Processing engine commands...")
+        commands = []
+
+        for cmd_name, data in engine.commands.iteritems():
+            engine.logger.debug("Processing command: %s" % cmd_name)
+            props = data["properties"]
+            app = props.get("app")
+
+            if app:
+                app_name = app.name
+            else:
+                app_name = None
+
+            commands.append(
+                dict(
+                    name=cmd_name,
+                    title=props.get("title", cmd_name),
+                    deny_permissions=props.get("deny_permissions", []),
+                    supports_multiple_selection=props.get(
+                        "supports_multiple_selection",
+                        False
+                    ),
+                    app_name=app_name,
+                    group=props.get("group"),
+                    group_default=props.get("group_default"),
+                    engine_name=props.get("engine_name"),
+                ),
+            )
+
+        engine.logger.debug("Engine commands processed.")
+        engine.logger.debug("Inserting commands into cache...")
+
+        # Connect to the database and get the hashes we need to include in
+        # the insert. Each of the lookups call out to the browser_integration
+        # core hook.
+        connection = sqlite3.connect(cache_file)
+        connection.text_factory = str
+        cursor = connection.cursor()
+
+        try:
+            commands_blob = sqlite3.Binary(
+                cPickle.dumps(commands, cPickle.HIGHEST_PROTOCOL)
+            )
+
+            # Since we're likely to be updating out-of-date cached data more
+            # often than we're going to be inserting new rows into the cache,
+            # we'll try an update first. If no rows were affected by the update,
+            # we move on to an insert.
+            cursor.execute(
```

Comment:
I tried and it wasn't working properly, but I'll try again. I got frustrated with the lack of a genuine upsert command in sqlite, and stackoverflows I found pointed to a couple things that didn't actually work. I said screw it and went with the approach here just because I knew it would work.